server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -213,20 +270,80 @@ remoteSchemaInterface schemaDoc defn@(G.InterfaceTypeDefinition description name
+          selSetFieldsWithoutFragments = (fmap . fmap) catMaybes selSetFieldsWithoutFragments'
+
+          -- common interface fields that exist in every
+          -- selection set provided
+          -- #1 of Note [Querying remote schema Interfaces]
+          commonInterfaceFields =
+            Map.elems $
+            Map.mapMaybe allTheSame $
+            Map.groupOn G._fName $
+            concatMap (\(_, flds) ->
+                         filter ((`elem` interfaceFieldNames) . G._fName) flds)
+                         $ selSetFieldsWithoutFragments
+
+          interfaceFieldNames = map G._fldName fields
+
+          allTheSame [] = Nothing
+          allTheSame (x:xs) = bool Nothing (Just x) $ all (== x) xs
+
+          -- #2 of Note [Querying remote schema interface fields]
+          nonCommonInterfaceFields =
+            catMaybes $ flip map selSetFieldsWithoutFragments $ \(objName, flds) ->
+              let nonCommonFields = filter (not . flip elem commonInterfaceFields) flds
+              in mkObjInlineFragment (objName, map G.SelectionField nonCommonFields)
+
+          -- helper function for #4 of Note [Querying remote schema interface fields]
+          mkObjInlineFragment (_, []) = Nothing
+          mkObjInlineFragment (objName, selSet) =
+            Just $ G.SelectionInlineFragment $
+              G.InlineFragment (Just objName) mempty selSet
+
+      -- #5 of Note [Querying remote schema interface fields]
+      in (map G.SelectionField commonInterfaceFields) <> nonCommonInterfaceFields
+
 -- | 'remoteSchemaUnion' returns a output parser for a given 'UnionTypeDefinition'.
 remoteSchemaUnion
   :: forall n m
   . (MonadSchema n m, MonadError QErr m)
   => SchemaIntrospection
   -> G.UnionTypeDefinition
-  -> m (Parser 'Output n ())
+  -> m (Parser 'Output n (SelectionSet NoFragments G.Name))
 remoteSchemaUnion schemaDoc defn@(G.UnionTypeDefinition description name _directives objectNames) =
   P.memoizeOn 'remoteSchemaObject defn do
-  objDefs <- traverse getObject objectNames
-  objs :: [Parser 'Output n ()] <- traverse (remoteSchemaObject schemaDoc) objDefs
+  objs :: [Parser 'Output n (Name, (SelectionSet NoFragments G.Name))] <-
+    traverse (getObjectParser schemaDoc getObject) objectNames
   when (null objs) $
     throw400 RemoteSchemaError $ "List of member types cannot be empty for union type " <> squote name
-  pure $ void $ P.selectionSetUnion name description objs
+  pure $ P.selectionSetUnion name description objs <&>
+    (\objNameAndSelSets ->
+       catMaybes $ flip map objNameAndSelSets $ \(objName, selSet) ->
+        case selSet of
+          -- In the case of an incoming query containing an union type, which has
+          -- 3 objects in it's object member types, but an incoming query only
+          -- queries 2 out of the 3 objects using fragments, then we get the
+          -- selectionSet for the object that was not queried as `[]`. So,
+          -- we omit it while constructing the query
```

Comment:
I had to read this a few times before I was convinced I got the point. Let me try writing this as well.

> The return value obtained from the parsing of a union selection set specifies, for each object in the union type, a fragment-free selection set for that object type. In particular, if, for a given object type, the selection set passed to the union type did not specify any fields for that object type (i.e. if no inline fragment applied to that object), the selection set resulting from the parsing through that object type would be empty, i.e. `[]`. We exclude such object types from the reconstructed selection set for the union type, as selection sets cannot be empty.

Reply:
Yes, that's right. Also, Your explanation is much better. I'll use your explanation in the code.